File: DeepSpace-Java/src/deepspace/EnemyStarShip.java
package deepspace;

class EnemyStarShip implements Copyable <EnemyStarShip>, SpaceFighter {

    private Loot loot;
    private Damage damage;
    private String name;
    private float ammoPower;
    private float shieldPower;

//
// ──────────────────────────────────────────────────────────── CONSTRUCTORES ─────
//

    EnemyStarShip (String n, float a, float s, Loot l, Damage d) {
        name        = n;
        loot        = l;
        damage      = d;
        ammoPower   = a;
        shieldPower = s;
    }

    EnemyStarShip (EnemyStarShip s) {
        name        = s.name;
        loot        = s.loot;
        damage      = s.damage;
        ammoPower   = s.ammoPower;
        shieldPower = s.shieldPower;
    }

    @Override
    public EnemyStarShip copy() {
        return new EnemyStarShip(this);
    }

    EnemyToUI getUIversion() {
        return new EnemyToUI(this);
    }

//
// ────────────────────────────────────────────────────────────── CONSULTORES ─────
//

    public float protection () {
        return shieldPower;
    }

    public float fire () {
        return ammoPower;
    }

    public float getAmmoPower() {
        return ammoPower;
    }

    public Damage getDamage() {
        return damage;
    }

    public Loot getLoot() {
        return loot;
    }

    public String getName() {
        return name;
    }

    public float getShieldPower() {
        return shieldPower;
    }

    public ShotResult receiveShot (float shot) {
        if (shot >= shieldPower)
            return ShotResult.DONOTRESIST;
        else
            return ShotResult.RESIST;
    }

    public String toString () {
        return    "-> Nombre: " + name
                + "\n\t-> Potencia: " + ammoPower
                + "\n\t-> Escudos: " + shieldPower
                + "\n\t-> Loot: " + loot.toString()
                + "\n\t-> Daño:" + damage.toString();

    }
}

File: DeepSpace-Ruby/lib/EnemyStarShip.rb
# encoding:utf-8
require_relative 'Loot'
require_relative 'Damage'
require_relative 'ShotResult'
require_relative 'EnemyToUI'

module Deepspace
class EnemyStarShip

#
# ──────────────────────────────────────────────────────────── CONSTRUCTORES ─────
#

    def initialize (n, a, s, l, d)
        @name        = n
        @ammoPower   = a
        @shieldPower = s
        @loot        = l
        @damage      = d
    end

    def self.newCopy(s)
        new(s.name, s.ammoPower, s.shieldPower, s.loot, s.damage)
    end


    def getUIversion()
        EnemyToUI.new(self)
    end

#
# ─────────────────────────────────────────────────────────────── INTERFACES ─────
#

    def protection
        @shieldPower
    end

    def fire
        @ammoPower
    end

    def receiveShot (shot)
        if shot >= @shieldPower
            ShotResult::DONOTRESIST
        else
            ShotResult::RESIST
        end
    end

    def to_s
        getUIversion.to_s
    end

    attr_reader :name, :loot, :damage, :ammoPower, :shieldPower
end
end
